### DISCUSSION TITLE
Heat source of Heat Conduction

### DISCUSSION THREAD

**Discussion by Salma-Mao** [idaholab/moose#20625] [Q&A General] [answered]
There are two block with different material properties. The heat source [FunctionPathEllipsoidHeatSource](https://mooseframework.inl.gov/source/materials/FunctionPathEllipsoidHeatSource.html) is applied on the surface of block 1. A strange phenomenon has occurred. The heat source is applied inside the block 1. However, when there is one block, the heat source is applied on the surface of the block. The centre of heat source of the two case is same. Why does this happen？
case 1 - just one block
![image](https://user-images.githubusercontent.com/54804946/159667658-d90d69e2-ef4f-40fb-8d91-6f9a24f0bd1b.png)

```
[Mesh]
  type = GeneratedMesh
  dim = 3
  xmin = 0
  xmax = 0.5
  nx = 50

  ymin = 0
  ymax = 0.3
  ny = 30

  zmin = 0.0
  zmax = 0.3
  nz = 50
[]
[Variables]
  [./temp]
    initial_condition = 300
  [../]
[]

[Kernels]
  [./time]
    type = ADHeatConductionTimeDerivative
    variable = temp
  [../]
  [./heat_conduct]
    type = ADHeatConduction
    variable = temp
    thermal_conductivity = thermal_conductivity
  [../]
  [./heat_source]
    type = ADMatHeatSource
    material_property = volumetric_heat
    variable = temp
  [../]
[]

[BCs]
  [./temp_bottom_fix]
    type = ADDirichletBC
    variable = temp
    boundary = 1
    value = 300
  [../] 
[]

[Materials]
  [./heat]
    type = ADHeatConductionMaterial
    specific_heat = 305
    thermal_conductivity = 10e-2
  [../]
  [./density]
    type = ADGenericConstantMaterial
    prop_names = 'density'
    prop_values = '6.35e-6'
  [../]
  [./volumetric_heat]
    type = FunctionPathEllipsoidHeatSource
    rx = 0.05
    ry = 0.05
    rz = 0.12
    power = 200
    efficiency = 0.36
    factor = 2
    function_x= path_x
    function_y= path_y
    function_z= path_z
  [../]
[]

[Functions]
  [./path_x]
    type = ParsedFunction
    value =  1200*t+0.08
  [../]
  [./path_y]
    type = ParsedFunction
    value =  0.15
  [../]
  [./path_z]
    type = ParsedFunction
    value = 0.3
  [../]
[]
```

case 2 - two blocks
![image](https://user-images.githubusercontent.com/54804946/159668250-344715a7-12d6-4fb4-8bf4-d63d402085f9.png)

```
[Mesh]

  [./block_1]
  type = GeneratedMeshGenerator
  dim = 3
  xmin = 0
  xmax = 0.5
  nx = 50

  ymin = 0
  ymax = 0.3
  ny = 30

  zmin =0
  zmax = 0.27
  nz = 27


[../]

[./block_1_sidesets]
    type = RenameBoundaryGenerator
    input = block_1
    old_boundary = '0 1 2 3 4 5'
    new_boundary = '6                  7              8                  9              10               11'
                  # substrate_back substrate_bottom substrate_right substrate_top substrate_left substrate_front

  [../]

  [./block_1_id]
    type = SubdomainIDGenerator
    input = block_1_sidesets
    subdomain_id = 1
  [../]

 [./block_2]
 type = GeneratedMeshGenerator
 dim = 3
 xmin = 0
 xmax = 0.5
 nx = 100

 ymin = 0
 ymax =0.3
 ny = 60

 zmin = 0.27
 zmax = 0.3
 nz = 6


[../]

[./block_2_sidesets]
    type = RenameBoundaryGenerator
    input = block_2
    old_boundary = '0 1 2 3 4 5'
    new_boundary = '100       101            102             103        104        105'

              

  [../]

[./block_2_id]
  type = SubdomainIDGenerator
  input = block_2_sidesets
  subdomain_id = 2
  [../]

[./combined]
    type = MeshCollectionGenerator
    inputs = 'block_1_id block_2_id'
    [../]



[]

[Variables]
  [./temp]
    initial_condition = 400
    block = '1 2'
  [../]

[]



[Kernels]
  [./time_powder]
    type = ADHeatConductionTimeDerivative
    variable = temp
    specific_heat = specific_heat
    density_name =density
    block = '1 2'
  []

  [./heat_conduct_powder]
    type = ADHeatConduction
    thermal_conductivity =  thermal_conductivity
    variable = temp
    block = '1 2'
  [../]

  [./heat_source_powder]
    type = ADMatHeatSource
    material_property = volumetric_heat
    variable = temp
      block = '1 2'
  [../]

[]

[BCs]
  [./temp_bottom_fix]
    type = ADDirichletBC
    variable = temp
    boundary ='104 102 105 100 6 8 10 11'

        
    value = 400
  [../]


  [./convention_surfcae]
      type = ADConvectiveHeatFluxBC
      variable = temp
      boundary = 103
      T_infinity = 300.0
      heat_transfer_coefficient = 110
    [../]

[]

[Materials]
[./heat_substrate]
  type = ADHeatConductionMaterial
  specific_heat = 470
  thermal_conductivity = 1.34e-2
  block = 1
[../]

[./density_substrate]
  type = ADGenericConstantMaterial
  prop_names = 'density'
  prop_values = '7.59e-6'
  block = 1
[../]

[./heat_powder]
  type = ADHeatConductionMaterial
  specific_heat = 235
  thermal_conductivity = 6.7e-3
  block = 2
[../]

[./density_powder]
  type = ADGenericConstantMaterial
  prop_names = 'density'
  prop_values = '3.975e-6'
  block = 2
[../]


  [./volumetric_heat]
    type = Function_PathEllipsoid_HeatSource
    rx = 0.05
    ry = 0.05
    rz = 0.12
    efficiency = 0.36
    power =150
    factor = 1
    function_x= path_x
    function_y= path_y
    function_z= path_z
  [../]
[]

[Functions]
  [./path_x]
    type = ParsedFunction
    value = 900*t+0.08
  [../]
  [./path_y]
    type = ParsedFunction
    value = 0.15
  [../]
  [./path_z]
    type = ParsedFunction
    value = 0.3
  [../]

[]
```

Question 2,
In the second case, there is no  heat transfer from block 1 to block 2. However, the Heat Conduction Module is applied on the two block. Is there lack of Thermal Contact Action?

Question 3,
The effective depth ellipsoid radius exceeds the thickness of the block. If the heat source is applied on the two blocks,  heat source is reloaded on the two block. Why the  heat source is applied on the two block, there is  no energy on the second block?

**Comment by GiudGiud** [comment, marked as answer]
Hello

You can use the block parameter, `block = 1` to restrict the heat source out of block 2. 

There is no heat transfer between block 1 and 2 because:
- the MeshCollectionGenerator does not stitch the two meshes together (as it says in the docs), they are separate. Use this instead:
https://mooseframework.inl.gov/source/meshgenerators/StitchedMeshGenerator.html

- I dont see any of the sidesets you tried to generate when I ran your input file. You should check what the initial names of the sidesets are before using a `RenameBoundaryGenerator`

Format wise:
- if pasting input files rather than linking to a github repo, please dont forget any blocks. Both these inputs are missing Executioners and Outputs.
- dont change names of objects: `Function_PathEllipsoid_HeatSource` wont be recognized
- Use triple quotes to format pasted input files

Guillaume

**Reply by wangzhaohao** [reply]
```
[Mesh]
    [./ccmh]
        type = ConcentricCircleMeshGenerator
        num_sectors = 20
        radii = '0.23E-2 1.805E-2 1.81E-2 1.86E-2'
        #        Zr-rod fuel He S.S
        rings = '1      5     1    5'
        has_outer_square = off #on
        # ptich = 1.42063
#       portion = left_half
        preserve_volumes = on
#        inner_mesh_fraction = 0.9#0.5
        smoothing_max_it = 3
    []
    [./HeOuter]
        type = SideSetsBetweenSubdomainsGenerator
        input = ccmh
        primary_block = 4
        paired_block = 3
        new_boundary = 'HeOuter'
    [../]
    [./HeInner]
        type = SideSetsAroundSubdomainGenerator
        input = HeOuter
        block = 2
        new_boundary = 'HeInner'
    [../]
    [./ed0]
        type = BlockDeletionGenerator
        input = HeInner
        block = 3
    [../]
    [./extrude]
        type = FancyExtruderGenerator
        input = ed0
        heights = '130E-3 260E-3'
        num_layers = '1 2'
        direction = '0 0 1'
        top_boundary = 998
        bottom_boundary = 999
        subdomain_swaps = '1 1  2 2 4 4;
                           1 11 2 12 4 14'
    [../]
[]

[Problem]
    coord_type = XYZ
#    rz_coord_axis = Z
[]

[GlobalParams]
    order = FIRST
[]

[Variables]
    [temp]
        order = FIRST
        family = LAGRANGE
        initial_condition = 300.0
    []
[]

[AuxVariables]
    [./gap_conductance]
        order = CONSTANT
        family = MONOMIAL
    [../]
[]

[Kernels]
    [./heat]
        type = HeatConduction
        variable = temp
        block = '1 11 2 12 4 14'
    [../]
    [./heat_source]
        type = HeatSource
        value = 6.6E7
        variable = temp
        block = '2 12'
    [../]
    [./heat_derivative]
        type = HeatConductionTimeDerivative
        variable = temp
    [../]
[]

[AuxKernels]
    [./gap_cond]
        type = MaterialRealAux
        property = gap_conductance
        variable = gap_conductance
        boundary = HeOuter
    [../]
[]

[BCs]
    [temp_clad]
        type = DirichletBC
        boundary = outer
        variable = temp
        value = 400.0
    []
[]

[Materials]
    [heat]
        type = HeatConductionMaterial
        block = '1 11 2 12'
        specific_heat = 299.96
        thermal_conductivity = 17
    []
    [heat1]
        type = HeatConductionMaterial
        block = '4 14'
        specific_heat = 400.0
        thermal_conductivity = 10
    []
    [density]
        type = GenericConstantMaterial
        prop_names = density
        prop_values = 6.161E3
    []
[]

[ThermalContact]
    [thermal_contact]
        type = GapHeatTransfer
        variable = temp
        primary = HeInner
        secondary = HeOuter
        emissivity_primary = 0
        emissivity_secondary = 0
        gap_conductivity = 5678
        quadrature = true
        gap_geometry_type = CYLINDER
        cylinder_axis_point_1 = '0 0 0'
        cylinder_axis_point_2 = '0 0 1'
    []
[]

[Executioner]
    type = Transient
    solve_type = 'PJFNK'
    
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
    petsc_options_value = 'lu        superlu_dist'

    start_time = 0.0
    end_time = 1.0
    dt = 0.1
    dtmin = 0.0001

    nl_rel_tol = 1E-12
    nl_abs_tol = 1E-7

    [./Quadrature]
        order = fifth
        side_order = seventh
    [../]
[]

[Outputs]
    exodus = true
[]
```
hello, there is some problems in the HeatSource, when i use this code. i find the temp is not right. 
![image](https://user-images.githubusercontent.com/51054734/160031083-e2f42d05-527f-4bd2-940f-1d3e546cb49f.png)
just like the block 2 12 is not exist.

**Reply by GiudGiud** [reply]
@wangzhaohao please create your own post on the forum

The temperature is right, it's your setup that looks weird. You set an initial condition of 300 in the volume, then with the dirichlet boundary condition you set 400 on the boundary. So the volume slowly heats up from both the volumetric heat source and from the dirichlet BC.

If the equilibrium state doesnt look right please let us know, but 1s of simulation is not enough to reach it.

**Reply by wangzhaohao** [reply]
i think moose compute one step, and the initial is replacement, so i set that, but when i change it to Steady. the problem is also exist.
![image](https://user-images.githubusercontent.com/51054734/160040780-0ac20f1f-0d2e-45c3-ba48-4dbe82d9ed06.png)
![image](https://user-images.githubusercontent.com/51054734/160040850-84a7b843-87e8-4843-9842-409c23a3ec4f.png)

**Reply by wangzhaohao** [reply]
sorry i don't create my own forum

**Reply by GiudGiud** [reply]
The problem is that HeInner is both around the annular region and inside the annular fuel. You should use this generator to generate that boundary, not the sidesetaroundsubdomaingenerator
```
    [./HeInner]
        type = SideSetsBetweenSubdomainsGenerator
        input = HeOuter
        primary_block = 2
        paired_block = 3
        new_boundary = 'HeInner'
    [../]
```

I encourage you to click the advanced (cog button in your screenshot) button in properties and look at the sideset and nodeset definitions.

**Reply by wangzhaohao** [reply]
thanks a lot, i use the SideSetBetweenSubdomainsGenerator replace SideSetAroundSubdomainsGenerator.

**Reply by Salma-Mao** [reply]
> Hello
> 
> You can use the block parameter, `block = 1` to restrict the heat source out of block 2.
> 
> There is no heat transfer between block 1 and 2 because:
> 
> * the MeshCollectionGenerator does not stitch the two meshes together (as it says in the docs), they are separate. Use this instead:
>   https://mooseframework.inl.gov/source/meshgenerators/StitchedMeshGenerator.html
> * I dont see any of the sidesets you tried to generate when I ran your input file. You should check what the initial names of the sidesets are before using a `RenameBoundaryGenerator`
> 
> Format wise:
> 
> * if pasting input files rather than linking to a github repo, please dont forget any blocks. Both these inputs are missing Executioners and Outputs.
> * dont change names of objects: `Function_PathEllipsoid_HeatSource` wont be recognized
> * Use triple quotes to format pasted input files
> 
> Guillaume


Hello, there are two question.

Q1

The StitchedMeshGenerator is used in the case 2. There is heat transfer from block 1 to block 2. However, it is very strange that the most of the heat accumulates in block 1. A small amount of heat is transferred to block 2. In others' research, the highest temperature in the laser heating process was 3000 K in the same condition, while my temperature exceed a lot. The case is shown below.

![image](https://user-images.githubusercontent.com/54804946/161228275-55c9dd4b-65de-4b15-8468-8f4c7463a1d2.png)


Q2

After I changed the code of case, its output became wrong. When Docker was restarted, the results were still wrong. However, I create a new case with the modified code, and the output is ideal. I want to why this is happening.The example is that the previous case 2 and the below.
```
[Mesh]
[./substrate]
type = GeneratedMeshGenerator
dim = 3
xmin = 0
xmax =0.5
nx = 50

ymin = 0
ymax = 0.3
ny = 30

zmin =0
zmax = 0.3
nz = 30
[../]
[./renameboundary_substrate]
type = RenameBoundaryGenerator
input = substrate
old_boundary = '0 1 2 3 4 5'
new_boundary = 'substrate_back substrate_bottom substrate_right substrate_top substrate_left substrate_front'
[../]
[./substrate_id]
type = SubdomainIDGenerator
input = renameboundary_substrate
subdomain_id = 1
[../]
[./powder]
type = GeneratedMeshGenerator
dim = 3
xmin = 0
xmax = 0.5
nx = 50

ymin = 0
ymax = 0.3
ny = 50

zmin =0.3
zmax = 0.33
nz = 6
[../]
[./renameboundary_powder]
type = RenameBoundaryGenerator
input = powder
old_boundary = '0 1 2 3 4 5'
new_boundary = '100 101 102 103 104 105'
[../]
[./renameboundary_powder_rename]
type = RenameBoundaryGenerator
input = renameboundary_powder
old_boundary = '100 101 102 103 104 105'
new_boundary = 'powder_back powder_bottom powdere_right powder_top powder_left powder_front'
[../]
[./powder_id]
type = SubdomainIDGenerator
input = renameboundary_powder_rename
subdomain_id = 2
[../]
[./combined_mesh]
type = StitchedMeshGenerator
inputs = 'substrate_id powder_id'
stitch_boundaries_pairs = 'substrate_top powder_bottom '
[../]

[]

[Variables]
[./temp]
initial_condition = 300
[../]
[]

[Kernels]
[./time]
type = ADHeatConductionTimeDerivative
variable = temp
block = '1 2'
[../]
[./heat_conduct]
type = ADHeatConduction
variable = temp
thermal_conductivity = thermal_conductivity
block = '1 2'
[../]
[./heat_source]
type = ADMatHeatSource
material_property = volumetric_heat
variable = temp
block = 2
[../]
[]

[ThermalContact]
[thermal_contact]
type = GapHeatTransfer
variable = temp
primary = powder_bottom
secondary = substrate_top
emissivity_primary = 0
emissivity_secondary = 0
gap_conductivity = 1.0e6
[]
[]

[BCs]
[./temp_bottom_fix]
type = ADDirichletBC
variable = temp
boundary = 1
value = 300
[../]
[]

[Materials]
[./heat_2]
type = ADHeatConductionMaterial
specific_heat = 470.0
thermal_conductivity = 4.0875e-4
block = 2
[../]
[./density_2]
type = ADGenericConstantMaterial
prop_names = 'density'
prop_values = '3.796e-6'
block = 2
[../]

[./heat_1]
type = ADHeatConductionMaterial
specific_heat = 470.0
thermal_conductivity = 1.34e-2
block = 1
[../]
[./density_1]
type = ADGenericConstantMaterial
prop_names = 'density'
prop_values = '7.592e-6'
block = 1
[../]

[./volumetric_heat]
type = FunctionPathEllipsoidHeatSource
rx = 0.05
ry = 0.05
rz = 0.12
power = 200
efficiency = 0.36
factor = 1
function_x= path_x
function_y= path_y
function_z= path_z
[../]
[]

[Functions]
[./path_x]
type = ParsedFunction
value = 900*t+0.08
[../]
[./path_y]
type = ParsedFunction
value = 0.15
[../]
[./path_z]
type = ParsedFunction
value =0.33
[../]
[]

[Postprocessors]
[temp_max]
type = ElementExtremeValue
variable = temp
[]
[temp_min]
type = ElementExtremeValue
variable = temp
value_type = min
[]
[temp_avg]
type = ElementAverageValue
variable = temp
[]
[]

[Preconditioning]
[./full]
type = SMP
full = true
[../]
[]

[Executioner]
type = Transient
solve_type = PJFNK

nl_rel_tol = 1e-6
nl_abs_tol = 1e-6

petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
petsc_options_value = 'preonly lu superlu_dist'

l_max_its = 100

end_time =3e-4
dt = 1.0e-7
dtmin = 1.0e-9
[]

[Outputs]
exodus = true
csv = true
[]
```

**Reply by GiudGiud** [reply]
Hello

The input file must have changed, MOOSE behaves deterministically and the objects you are using have not changed in the meantime. 
Can you try to `diff input1.i input2.i` in the console and see what's different?
PS: It's triple back-quotes to format code snippets, not single back quotes. 

Guillaume

**Reply by Salma-Mao** [reply]
Hello
The input1.i 
、、、
[Mesh]

  [./block_1]
  type = GeneratedMeshGenerator
  dim = 3
  xmin = 0
  xmax = 0.5
  nx = 50

  ymin = 0
  ymax = 0.3
  ny = 30

  zmin =0
  zmax = 0.3
  nz = 30


[../]

[./block_1_sidesets]
    type = RenameBoundaryGenerator
    input = block_1
    old_boundary = '0 1 2 3 4 5'
    new_boundary = '6                  7              8                  9              10               11'
                  # substrate_back substrate_bottom substrate_right substrate_top substrate_left substrate_front

  [../]

  [./block_1_id]
    type = SubdomainIDGenerator
    input = block_1_sidesets
    subdomain_id = 1
  [../]

 [./block_2]
 type = GeneratedMeshGenerator
 dim = 3
 xmin = 0
 xmax = 0.5
 nx = 100

 ymin = 0
 ymax =0.3
 ny = 60

 zmin = 0.3
 zmax = 0.33
 nz = 6


[../]

[./block_2_sidesets]
    type = RenameBoundaryGenerator
    input = block_2
    old_boundary = '0 1 2 3 4 5'
    new_boundary = '100       101            102             103        104        105'

              

  [../]

[./block_2_id]
  type = SubdomainIDGenerator
  input = block_2_sidesets
  subdomain_id = 2
  [../]

[./combined]
    type = MeshCollectionGenerator
    inputs = 'block_1_id block_2_id'
    [../]



[]

[Variables]
  [./temp]
    initial_condition = 400
    block = '1 2'
  [../]

[]



[Kernels]
  [./time_powder]
    type = ADHeatConductionTimeDerivative
    variable = temp
    specific_heat = specific_heat
    density_name =density
    block = '1 2'
  []

  [./heat_conduct_powder]
    type = ADHeatConduction
    thermal_conductivity =  thermal_conductivity
    variable = temp
    block = '1 2'
  [../]

  [./heat_source_powder]
    type = ADMatHeatSource
    material_property = volumetric_heat
    variable = temp
      block = '1 2'
  [../]

[]

[BCs]
  [./temp_bottom_fix]
    type = ADDirichletBC
    variable = temp
    boundary ='104 102 105 100 6 8 10 11'

        
    value = 400
  [../]


  [./convention_surfcae]
      type = ADConvectiveHeatFluxBC
      variable = temp
      boundary = 103
      T_infinity = 300.0
      heat_transfer_coefficient = 110
    [../]

[]

[Materials]
[./heat_substrate]
  type = ADHeatConductionMaterial
  specific_heat = 470
  thermal_conductivity = 1.34e-2
  block = 1
[../]

[./density_substrate]
  type = ADGenericConstantMaterial
  prop_names = 'density'
  prop_values = '7.59e-6'
  block = 1
[../]

[./heat_powder]
  type = ADHeatConductionMaterial
  specific_heat = 235
  thermal_conductivity = 6.7e-3
  block = 2
[../]

[./density_powder]
  type = ADGenericConstantMaterial
  prop_names = 'density'
  prop_values = '3.975e-6'
  block = 2
[../]


  [./volumetric_heat]
    type = Function_PathEllipsoid_HeatSource
    rx = 0.05
    ry = 0.05
    rz = 0.12
    efficiency = 0.36
    power =150
    factor = 1
    function_x= path_x
    function_y= path_y
    function_z= path_z
  [../]
[]

[Functions]
  [./path_x]
    type = ParsedFunction
    value = 900*t+0.08
  [../]
  [./path_y]
    type = ParsedFunction
    value = 0.15
  [../]
  [./path_z]
    type = ParsedFunction
    value = 0.33
  [../]

[]

[Postprocessors]
[temp_max]
type = ElementExtremeValue
variable = temp
[]
[temp_min]
type = ElementExtremeValue
variable = temp
value_type = min
[]
[temp_avg]
type = ElementAverageValue
variable = temp
[]
[]

[Preconditioning]
[./full]
type = SMP
full = true
[../]
[]

[Executioner]
type = Transient
solve_type = PJFNK

nl_rel_tol = 1e-6
nl_abs_tol = 1e-6

petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
petsc_options_value = 'preonly lu superlu_dist'

l_max_its = 100

end_time =3e-4
dt = 1.0e-7
dtmin = 1.0e-9
[]

[Outputs]
exodus = true
csv = true
[]


、、、

The input2.i
、、、
[Mesh]
[./substrate]
type = GeneratedMeshGenerator
dim = 3
xmin = 0
xmax =0.5
nx = 50

ymin = 0
ymax = 0.3
ny = 30

zmin =0
zmax = 0.3
nz = 30
[../]
[./renameboundary_substrate]
type = RenameBoundaryGenerator
input = substrate
old_boundary = '0 1 2 3 4 5'
new_boundary = 'substrate_back substrate_bottom substrate_right substrate_top substrate_left substrate_front'
[../]
[./substrate_id]
type = SubdomainIDGenerator
input = renameboundary_substrate
subdomain_id = 1
[../]
[./powder]
type = GeneratedMeshGenerator
dim = 3
xmin = 0
xmax = 0.5
nx = 50

ymin = 0
ymax = 0.3
ny = 50

zmin =0.3
zmax = 0.33
nz = 6
[../]
[./renameboundary_powder]
type = RenameBoundaryGenerator
input = powder
old_boundary = '0 1 2 3 4 5'
new_boundary = '100 101 102 103 104 105'
[../]
[./renameboundary_powder_rename]
type = RenameBoundaryGenerator
input = renameboundary_powder
old_boundary = '100 101 102 103 104 105'
new_boundary = 'powder_back powder_bottom powdere_right powder_top powder_left powder_front'
[../]
[./powder_id]
type = SubdomainIDGenerator
input = renameboundary_powder_rename
subdomain_id = 2
[../]
[./combined_mesh]
type = StitchedMeshGenerator
inputs = 'substrate_id powder_id'
stitch_boundaries_pairs = 'substrate_top powder_bottom '
[../]

[]

[Variables]
[./temp]
initial_condition = 300
[../]
[]

[Kernels]
[./time]
type = ADHeatConductionTimeDerivative
variable = temp
block = '1 2'
[../]
[./heat_conduct]
type = ADHeatConduction
variable = temp
thermal_conductivity = thermal_conductivity
block = '1 2'
[../]
[./heat_source]
type = ADMatHeatSource
material_property = volumetric_heat
variable = temp
block = 2
[../]
[]

[ThermalContact]
[thermal_contact]
type = GapHeatTransfer
variable = temp
primary = powder_bottom
secondary = substrate_top
emissivity_primary = 0
emissivity_secondary = 0
gap_conductivity = 1.0e6
[]
[]

[BCs]
[./temp_bottom_fix]
type = ADDirichletBC
variable = temp
boundary = 1
value = 300
[../]
[]

[Materials]
[./heat_2]
type = ADHeatConductionMaterial
specific_heat = 470.0
thermal_conductivity = 4.0875e-4
block = 2
[../]
[./density_2]
type = ADGenericConstantMaterial
prop_names = 'density'
prop_values = '3.796e-6'
block = 2
[../]

[./heat_1]
type = ADHeatConductionMaterial
specific_heat = 470.0
thermal_conductivity = 1.34e-2
block = 1
[../]
[./density_1]
type = ADGenericConstantMaterial
prop_names = 'density'
prop_values = '7.592e-6'
block = 1
[../]

[./volumetric_heat]
type = FunctionPathEllipsoidHeatSource
rx = 0.05
ry = 0.05
rz = 0.12
power = 200
efficiency = 0.36
factor = 1
function_x= path_x
function_y= path_y
function_z= path_z
[../]
[]

[Functions]
[./path_x]
type = ParsedFunction
value = 900*t+0.08
[../]
[./path_y]
type = ParsedFunction
value = 0.15
[../]
[./path_z]
type = ParsedFunction
value =0.33
[../]
[]

[Postprocessors]
[temp_max]
type = ElementExtremeValue
variable = temp
[]
[temp_min]
type = ElementExtremeValue
variable = temp
value_type = min
[]
[temp_avg]
type = ElementAverageValue
variable = temp
[]
[]

[Preconditioning]
[./full]
type = SMP
full = true
[../]
[]

[Executioner]
type = Transient
solve_type = PJFNK

nl_rel_tol = 1e-6
nl_abs_tol = 1e-6

petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
petsc_options_value = 'preonly lu superlu_dist'

l_max_its = 100

end_time =3e-4
dt = 1.0e-7
dtmin = 1.0e-9
[]

[Outputs]
exodus = true
csv = true
[]
、、、


The two case get the different output.

The case 1
![image](https://user-images.githubusercontent.com/54804946/161362924-316d5281-36ef-44e0-afd6-5c9a41a958af.png)

The case 2

![image](https://user-images.githubusercontent.com/54804946/161362948-113dd85f-a81f-467c-8f5b-8dd41985061d.png)


You can find that the centre of heat source defined in the case is same. However, the output shown that the centre of heat source is different.

Q2
There is heat transfer from block 1 to block 2. However, it is very strange that the most of the heat accumulates in block 1. A small amount of heat is transferred to block 2. My classmate simulated the same case with ANSYS and got the following output.  You can see that the heat is very well transferred from block 1 to block 2.  

My case (the input2.i)
![image](https://user-images.githubusercontent.com/54804946/161363326-036c37fd-3c17-4d59-a335-55a7aa5dd9fb.png)

My classmate's case
![image](https://user-images.githubusercontent.com/54804946/161363233-bfa23b40-6c8b-4ba3-8fd3-c742a2cd7624.png)

Thanks
MS

**Reply by Salma-Mao** [reply]
Thanks for your  help. I still don't  find what makes the different. But  I can create new case to deal with the problems. The reason of  heat   transfer  between two block  is that the grid size is different. 
Thanks again
MS

**Reply by GiudGiud** [reply]
Yeah the two mesh parts need to be stitched together, so they cant have different grid sizes. 
This is basically what I said here :)

> There is no heat transfer between block 1 and 2 because:
> 
> * the MeshCollectionGenerator does not stitch the two meshes together (as it says in the docs), they are separate. Use this instead:
>   https://mooseframework.inl.gov/source/meshgenerators/StitchedMeshGenerator.html
